### ISSUE TITLE
exists() method for repository

### ISSUE THREAD

**Issue by cunhazera** [quarkusio/quarkus#7884]
**Description**
Nowadays we have the `findByIdOptional` and `count` to check if a record exists inside the database. `count` is much faster than selecting the whole object data and serializing it. But wouldn't it be nice to have an `exists(id)` method to abstract a db query that is so common?

**Comment by gastaldi** [comment]
cc @FroMage @emmanuelbernard @gsmet

**Comment by FroMage** [comment]
Is it really that common to check for existence of a record without loading it? I've never had to do that myself.

**Comment by cunhazera** [comment]
@FroMage but isn't it too much to load all fields from db to check if it exists?

**Comment by FroMage** [comment]
Well, sure, but only if you need to check for existence without later using the entity. I've never seen such a use case. All my use-cases check if the object exist and then proceed to use the entity.

**Comment by cunhazera** [comment]
@FroMage yeah, it makes sense. In my case, I check the record exists. If it does, I throw a custom exception.

**Comment by Postremus** [comment]
@FroMage I also sometimes just want to check for existence, and throw an NotFoundException.

If you want, I can try to implement this feature.

**Comment by FroMage** [comment]
> and then throw an NotFoundException

But if it exists, what does it do?

**Comment by alexander-dammeier** [comment]
I use "exists" to prove if an id really refers to an existing aggregate when i only hold the id in another aggregate. It is a business rule in our system so i have to check it but i do not need the hole entity, only the already given id.

**Comment by emmanuelbernard** [comment]
note that count is not too far from what you have in exists.

if ( !Person.exists(1L) ) { ... }
// vs
if ( Person.count("id", 1L) != 1 ) { ... }

**Comment by alexander-dammeier** [comment]
But is exists faster?
Are all databases intelligent enough not to go through all rows to count duplicates in a column with an unique constraint?

**Comment by emmanuelbernard** [comment]
What do you want for the feature? The existence of the entity? Or the existence of some column value?
I thought the former based on the issue description. In this case, I don't think my count approach is any different than whatever a custom exist implementation would do performance wise. If anything it could be made more memory efficient than a `select id from entity where id = ?`. The count is done *after* the filtering.

The one annoyance with my solution is that the user must know that the id property is named "id"

**Comment by lordofthejars** [comment]
:+1: to implement this, I think that of course you can do a `count` but this `exists` methods is more about clean code. At the end you can always do in Java `list.size() == 0` but we do `list.isEmpty()` this is more or less the same.

**Comment by abuinitski** [comment]
`count` is not the same as `exists`.
Postgres example: 
- `SELECT EXISTS(SELECT 1 FROM contact WHERE name='John')` vs
- `SELECT COUNT(*) FROM contact WHERE name = 'John'`

Now imagine that name has no index on it (on purpose), and request results in a full-table-scan. Count will have to actually scan the whole table, while `EXISTS` can actually stop as soon as it sees first "John" guy.

**Comment by abuinitski** [comment]
> Sure, but the question remains: what do you do once you know the entity exists?

The question is not what "to do", the question is what "not to do".

For example:
1. Do not create a duplicate. 
2. Do not bother with custom native query and "ON CONFLICT" clause in case a duplicate already exists
3. Do not bother parsing native database exception code, and checking failing constraint name, to catch duplicate error and respond to client with a reasonable error message instead of an "unknown error something is wrong with my database idk"

Of course, race condition is still possible, but then I just can refer to low salary and bad management to justify why my code has issues.

**Comment by alexander-dammeier** [comment]
In our case, we often check for the existence of an entry because the existence alone tells us if a step in our process was executed successful. Otherwise we schedule a retry, so for this case we use it more like an event store and the entries are often immutable.

I think it is very dangerous to use an exist function to avoid duplicates as another process could create such an entry after you just checked it. In my opinion, unique constraints (ID conflicts) are the only viable option to truly avoid this and therefore we always check if the transaction failed because of a conflict. We then do not load the entry because knowing that it exists is enough to continue or there is simply nothing left to do.

**Comment by markus-lehr-dedalus** [comment]
Our company has a microservice architecture which also uses Kafka for communication.
Since Kafka leaves the fitering of messages to the clients, our app needs to decide if the received event is relevant for us.
For this I want to check if a used identifier in a Kafka message `exists` in our db (which contains millions of records, only a small subset of the billions of possible entities of our whole system).
Having to use a `count` instead is such a dealbreaker for us that I'll have to write a non-panache query instead.

**Comment by FroMage** [comment]
@gavinking is there an equivalent `.exists(Object id)` method in the latest ORM these days? I could not find anything in the `Session` API.

**Comment by gavinking** [comment]
No, and what would the semantics of such an operation be?

1. Is it enough to check the second-level cache, or does it have to always go to the database?
2. Does it acquire some sort of exclusive lock on that id so that no other transaction can insert or delete the row until the present transaction commits?

**Comment by gavinking** [comment]
But the main reason we don't have it is it's typically such a crappy and inefficient way to interact with the database.

**Comment by gavinking** [comment]
Probably the people who want this should be using `upsert()` or something else which doesn't suck.

**Comment by gavinking** [comment]
So, to be clear (I was posting from my phone). The people posting above are giving what looks to me like terrible advice on how to implement an upsert on databases which don't support the SQL `merge` statement.

The approach Hibernate uses to emulate `merge` on databases which don't have it, is the following:

1. attempt an `UPDATE` on the row, and check the returned rowcount (if it is 1, stop)
2. if it is 0, attempt an `INSERT`.

An alternative approach is to attempt the `INSERT` first, and have fancy code for checking if a `SQLException` represents a primary key violation.

The following is _not_ a good solution:

1. attempt to `SELECT` the row
2. if it exists, attempt an `UPDATE`, or otherwise,
3. if not, attempt an `INSERT` and 🙏🙏that it wasn't inserted since the `SELECT`, since we don't have any sort of lock on the nonexistent row.

Anyway, people don't need to be implementing this stuff themselves, since it's already implemented by `StatelessSession.upsert()`, which will automatically use a `MERGE` on platforms which support it.

**Comment by gavinking** [comment]
@alexander-dammeier

> I use "exists" to prove if an id really refers to an existing aggregate when i only hold the id in another aggregate. It is a business rule in our system so i have to check it but i do not need the hole entity, only the already given id.

And this is **terrible**.

- First of all: "aggregates" are not the right way to use JPA. You can read what we have to say about aggregates here: https://docs.jboss.org/hibernate/orm/7.0/introduction/html_single/Hibernate_Introduction.html#associations
- Second of all: referential integrity is the responsibility of the database! You should have foreign key constraints, not rubbish `select` queries to validate referential integrity.

I'm on my knees begging everyone to put down the DDD Jonestown Kool-aid and just try using JPA and your relational database in the normal way that these things were intended to be used.

**Comment by markus-lehr-dedalus** [comment]
Again, in our distributed system (hospital), we receive a lot of Kafka messages - only a small part of which is relevant for our app.
Whether or not a message is relevant is only decided if the identifiers match some others in our DB.
If the identifiers don't match, we don't want to store an entity but ignore the message. 
This is the expected case and there we don't want to upsert or merge or persist, but need to efficiently and quickly check if this DB entry exists. The events which we actually need to store are rare so an extra round trip is not the end of the world.

@gavinking - I see your point and I also read that others just need an upsert, but I'd like to believe that we actually have a legitimate usecase for this exists check.

**Comment by FroMage** [comment]
Yeah, agreed. I was sceptical because I assumed `exists` would always be followed by a `load`. @gavinking assumes it's for `upsert`. But @markus-lehr-dedalus was the first to raise a use-case where `exists` is not followed by either `load` or `update` but for other actions (perhaps not even on the DB: on other systems?).

I am not sure if it's a common use-case, though. But at least it's new to me :)

**Comment by gavinking** [comment]
@markus-lehr-dedalus In your use case, what are the transactional semantics of that?

**Comment by gavinking** [comment]
Another comment: in the case where `exists()` takes an id, I'm not clear on why a count query (for example, `getResultCount()` in Hibernate) is no good. I would have expected the query planner to be able to come up with something very efficient when it's the PK.

(Unclear to me why `exists` around a subquery would be better.)

**Comment by markus-lehr-dedalus** [comment]
@gavinking I'm not 100% certain what you mean by transactional semantics, but I can give you a workflow:

One of the specific workflows (simplified a bit) for us is that we get the Kafka message for any medical case update (`@Incoming` annotation on method, not `@Channel` on mutiny stuff).

After a bit of validation, we enter a method marked as `@Transactional` for processing the medical case.
This medical case has a number (the thing I called identifier, but not actually the DB id, not even unique constraint) and we check if our app has any entities (can be multiple as well, e.g. blood samples) which reference this medical case via this number (ideally with this `exists` method).
Since we don't store whole the medical case itself (huge, complex object and another app's responsibility), this number does not have any foreign key constraint in our app. I therefore don't think any optimization in `getResultCount` can be performed.

If we find any entity, then we know that we have to process the message's content further and create or update another whole different entity (physical location relevant for nurses) than the one we executed the `exists` check for.
After that, we send SSE updates to connected clients and then the transactional method ends.

If we don't find any entity (e.g. patient came in for something unrelated to blood samples), then we don't create any entity and the transactional method ends.

**Comment by gavinking** [comment]
And then you retry later?

**Comment by markus-lehr-dedalus** [comment]
No we don't retry later - the producers of the Kafka messages write to general purpose topics and consumers can subscribe to those changes. Imagine in the hospital someone shouting "Someone stubbed their toe!", but we might not be interested in this event. Whether or not we are interested - again - is written in our DB and can change at runtime.

You probably mean: If such an entity (which indicates we might have been interested in a medical case update) is created at a later point, then we must have missed those updates? Then we fetch the current information for the nurses via REST in a separate mechanism. See my attached impeccable ms paint diagram for clarity.

![Image](https://github.com/user-attachments/assets/0627918b-64ae-4704-a3c7-da9b2af90475)

**Comment by gavinking** [comment]
So, look, since we already opened the door by adding `SelectionQuery.getResultCount()`, we could in principle add `SelectionQuery.hasResults()` which you could use as follows:

```java
boolean exists = 
        session.createSelectionQuery("where key = ?1", Thing.class)
                .setParameter(1, key)
                .hasResults();
```

And the resulting SQL would be:

```sql
select 1 from things where key = ? limit 1
```

But honestly I'm just not clear on how generally-useful this would be. I mean, you can already write:

```java
boolean exists = 
        sesssion.createQuery("select 1 from Thing where key = ?1 limit 1", int.class)
                .getSingleResultOrNull() != null;
```
which is really not much worse.

@FroMage ?

**Comment by markus-lehr-dedalus** [comment]
@gavinking I like your proposed solution because of the enhanced expressiveness - however, I totally understand that not a lot of people will use this.
Of course, I am not blocked by this in any way, because I came up with quite a similar sql statement myself.

**Comment by FroMage** [comment]
For Panache, this would be:

```java
// assuming we're talking about a Panache query and this is not a primary key
boolean exists = Thing.exists("key", key);
```

Or for a primary key:

```java
boolean exists = Thing.existsById(key);
```

I can trivially implement the first one (query) by using the first example `.hasResults()`. I am not sure if `select 1 from Thing where key = ?1 limit 1` is the best implementation for all databases/queries. Is it the most optimal query? If yes, then I can also trivially implement the first one (query) with this (unless this falls into the same traps as counting which requires parsing the HQL and having multiple translations).

As for the second case (by id) the translation is not trivial, I think.

**Comment by gavinking** [comment]
> Is it the most optimal query?

Generalizing over all SQL databases is definitely above my pay grade.

I suppose the main alternative would be:

```sql
select exists(select 1 from Thing where key = ?1)
```

I just asked ChatGPT, and it claims that they are equivalent on Postgres, and recommends use of `limit` or equivalent on MySQL, Db2, and SQL Server. For Oracle it thinks you should use `rownum`. I'd take all that with a grain of salt, since ChatGPT can be very convincingly wrong on this sort of stuff.

**Comment by gavinking** [comment]
See https://hibernate.atlassian.net/browse/HHH-19541 and https://github.com/hibernate/hibernate-orm/pull/10334/files.

**Comment by FroMage** [comment]
I suppose it means I have to reopen this to support this new method :)

**Comment by handicraftsman** [comment]
> Is it really that common to check for existence of a record without loading it? I've never had to do that myself.

@FroMage yes it is, for example, in user registration form - when you want to check that this username is not occupied, you oftentimes do `select exists(select 1 from ... limit 1)`, because you don't care who is that user - you only care to know whether the username is used or not.

---

> Now imagine that name has no index on it (on purpose), and request results in a full-table-scan. Count will have to actually scan the whole table, while EXISTS can stop as soon as it sees first "John" guy.

As @abuinitski mentioned, there might be no index. While I cannot imagine such a situation that index is not present for a heavily used query, it's not a given that there will be an index. This is not really critical, as I *can* add the index, but still.

---

Additionally, I still don't care *semantically* about exact count of users - I just want the database to return true/false, and not the count to compare on the java code side.
